### ISSUE TITLE
bug(@vtmn/assets): wrong flag display on Firefox/Chrome

### ISSUE THREAD

**Issue by Jerome1337** [Decathlon/vitamin-web#1313]
**Describe the bug**

After a fresh upgrade of `@vtmn/assets` packages some flags are not correctly render on Firefox/Chrome. Flags not correctly renders are differents depending of the browser.

For example on Firefox:
<img width="142" alt="Screenshot 2022-11-24 at 2 15 02 PM" src="https://user-images.githubusercontent.com/7968649/203793816-dccd0b40-9318-4f4a-8c16-17c7a7f58354.png">
<img width="163" alt="Screenshot 2022-11-24 at 2 14 52 PM" src="https://user-images.githubusercontent.com/7968649/203793814-5ac107c9-e89c-4f4f-891c-35fa4bd5f156.png">

And Chrome:

<img width="180" alt="Screenshot 2022-11-24 at 2 15 11 PM" src="https://user-images.githubusercontent.com/7968649/203793948-0ef15661-0166-40c7-ab7e-44407a3b6d9b.png">
<img width="148" alt="Screenshot 2022-11-24 at 2 15 16 PM" src="https://user-images.githubusercontent.com/7968649/203793949-9a304a8c-f984-4862-a8c7-cbd941966899.png">

**Steps to reproduce**

Use the flag sprite and display a Poland flag on Firefox or Colombia flag on Chrome.

**Expected behavior**

Flags should be corrects on every browser.

**Browsers affected**

- Chrome
- Firefox
- Safari

**Version affected**

`@vtmn/assets: 0.3.0`

**Comment by thibault-mahe** [comment]
Hi @Jerome1337 , thanks for reaching us :) 
Could you be more specific on the way your are consuming the flags please ? Are you using an `img` tag ? the svg `use` ? etc. Maybe if you can provide us with a demo it could be easier to debug :) 

I try to use both the Poland and Colombia flag on my side (@vtmn/assets: 0.3.0 with NextJs) and it seems to work well on both Chrome and Firefox.

Thanks!

**Comment by Jerome1337** [comment]
Hi @thibault-mahe,

We use the Sprite like this

```js
import flags from '@vtmn/assets/dist/assets/sprite/assets.svg';

...

// item.label is for example France...
// item.value is for example fr or de...
<svg width="22" height="19" fill="#001018" key={item.value} id={`ChooseCountry-${item.label.toUpperCase()}`}>
  <use href={`${flags}#flag-${item.value}`} />
</svg>
```

I tried to clear browser cache, delete and reinstalling packages, downgrade `@vtmn/assets` package without success.
Problem is on our local env (without any cache) and pre-production/production envs (cached).

The only difference I find between problematic and good svg is the colored parts of flags are:

- in wrong one: filled with color paths
- in good one: paths with colored rects inside  

For example the Colombian flag (wrong one)

```svg
<symbol  viewBox='0 0 28 20' id='flag-co'>
 <g clip-path="url(#clip0_503_4477)"><rect width="28" height="20" rx="2" fill="white"></rect><mask id="mask0_503_4477" style="mask-type:alpha" maskunits="userSpaceOnUse" x="0" y="0" width="28" height="20"><rect width="28" height="20" rx="2" fill="white"></rect></mask><g mask="url(#mask0_503_4477)"><g filter="url(#filter0_d_503_4477)"><path fill-rule="evenodd" clip-rule="evenodd" d="M0 14.6667H28V9.33334H0V14.6667Z" fill="#0748AE"></path></g><g filter="url(#filter1_d_503_4477)"><path fill-rule="evenodd" clip-rule="evenodd" d="M0 20H28V14.6667H0V20Z" fill="#DE2035"></path></g><path fill-rule="evenodd" clip-rule="evenodd" d="M0 9.33333H28V0H0V9.33333Z" fill="#FFD935"></path></g></g><defs><filter id="filter0_d_503_4477" x="0" y="9.33334" width="28" height="5.33334" filterunits="userSpaceOnUse" color-interpolation-filters="sRGB"><feflood flood-opacity="0" result="BackgroundImageFix"><fecolormatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"><feoffset><fecolormatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.06 0"><feblend mode="normal" in2="BackgroundImageFix" result="effect1_dropShadow_503_4477"><feblend mode="normal" in="SourceGraphic" in2="effect1_dropShadow_503_4477" result="shape"></feblend></feblend></fecolormatrix></feoffset></fecolormatrix></feflood></filter><filter id="filter1_d_503_4477" x="0" y="14.6667" width="28" height="5.33334" filterunits="userSpaceOnUse" color-interpolation-filters="sRGB"><feflood flood-opacity="0" result="BackgroundImageFix"><fecolormatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"><feoffset><fecolormatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.06 0"><feblend mode="normal" in2="BackgroundImageFix" result="effect1_dropShadow_503_4477"><feblend mode="normal" in="SourceGraphic" in2="effect1_dropShadow_503_4477" result="shape"></feblend></feblend></fecolormatrix></feoffset></fecolormatrix></feflood></filter><clippath id="clip0_503_4477"><rect width="28" height="20" rx="2" fill="white"></rect></clippath></defs>
</symbol>
```

And the Belgian flag (good one)
```svg
<symbol  viewBox='0 0 28 20' id='flag-be'>
  <g clip-path="url(#clip0_503_3663)"><rect width="28" height="20" rx="2" fill="white"></rect><mask id="mask0_503_3663" style="mask-type:alpha" maskunits="userSpaceOnUse" x="0" y="0" width="28" height="20"><rect width="28" height="20" rx="2" fill="white"></rect></mask><g mask="url(#mask0_503_3663)"><rect x="13.3334" width="14.6667" height="20" fill="#FF4453"></rect><path fill-rule="evenodd" clip-rule="evenodd" d="M0 20H9.33333V0H0V20Z" fill="#262626"></path><path fill-rule="evenodd" clip-rule="evenodd" d="M9.33337 20H18.6667V0H9.33337V20Z" fill="#FFCF3C"></path></g></g><defs><clippath id="clip0_503_3663"><rect width="28" height="20" rx="2" fill="white"></rect></clippath></defs>
</symbol>
```

**Comment by thibault-mahe** [comment]
Thanks for all these helpful details @Jerome1337 

Here is some possible explanation : 

Some flag icons like the Colombia one are using local `url()` referencing for things like `clip-path`, `mask` or `filter`. The original references are often inside an svg `<defs>` tag, for instance:
<img width="886" alt="image" src="https://user-images.githubusercontent.com/5102622/204021494-3724dacb-647a-4ca9-aa8e-d7316de429a2.png">

When using the the svg `<use>` element from an external reference (in our case the sprite/assets.svg#some-flag), the original svg is cloned inside a shadow root to keep import in isolation. It seems that sometimes Chrome doesn't know how to handle a local fetch with `url()` inside a remote file and do not honor the original `defs` element.

Here is the shadow DOM of the Colombia flag not working in Chrome (with the sprite method):
<img width="483" alt="image" src="https://user-images.githubusercontent.com/5102622/204022234-ccba3dbc-d776-43e5-a632-2933444f4baa.png">

Here is the same icon working in Firefox:
<img width="572" alt="image" src="https://user-images.githubusercontent.com/5102622/204022387-2aedf2d5-9f17-49f4-83ae-9ebd972a31e8.png">

And finally here is the original symbol, inside the generated sprite/assets.svg, provided in both cases:
<img width="450" alt="image" src="https://user-images.githubusercontent.com/5102622/204022717-5eb3b982-c05c-45a4-9ab1-7e3be37fe5d1.png">

Note how Firefox succeeds in cloning the whole structure of the original svg (`symbol` and `defs`) while Chrome is only keeping the root svg group, loosing the `symbol` ref and the `defs`.

I will investigate a bit more on this issue, 2 fixes could be possible:
- make chrome respects the svg structure
- rework the issued icons/flags with simpler structure

Can you tell me how urgent this issue is? Do you need a quick fix or is it blocking? Can you use another strategy in the meantime, for example with an `img` tag?

Thank you!

**Comment by Jerome1337** [comment]
Note that on Firefox other flags a misrendered using the sprite method, for example the Bulgarian one.

If you want some help or need tests, feel free to ask me 😄

**Comment by thibault-mahe** [comment]
Hi @Jerome1337 

I spent several hours trying to find a solution today. I'm still in the same mess, and it's getting weirder and weirder^^

Given our 2 flags from Colombia and Poland : 
- Colombia is using <defs> refrencing clip-path, mask and filter
![image](https://user-images.githubusercontent.com/5102622/204285928-5b7ae8c1-f970-469f-b1a3-6db2e0fd0969.png)

- Poland is using <defs> referencing clip-path and mask
![image](https://user-images.githubusercontent.com/5102622/204286128-280969b0-15d7-43f0-b41c-aac50c2e7f2b.png)



Well, with the current configuration, the Poland flag is fine in Chrome, not in Firefox ; the Columbia flag is fine in Firefox, not in Chrome : 
<img width="183" alt="image" src="https://user-images.githubusercontent.com/5102622/204288090-7df9bc62-69fc-4d24-b51b-b2c3a33ab94b.png">
![image](https://user-images.githubusercontent.com/5102622/204288173-19c5a7d4-3126-4d94-a120-8708313b3ef5.png)

Great, that doesn't make any sense.

Now, investigating a bit more, if I remove all the mask, clip-path and filter from the Columbia flag, it seems that all this is useless and suddenly everything works magically on both browsers : 
![image](https://user-images.githubusercontent.com/5102622/204289159-b48d7700-4f74-4ccc-94a2-a490d953f3ae.png)

![image](https://user-images.githubusercontent.com/5102622/204289226-30f70446-ca57-4fab-9722-6c8e8d8f4d6f.png)

<img width="171" alt="image" src="https://user-images.githubusercontent.com/5102622/204289289-e16e52d0-6471-4a2d-a3c9-ded1302b28ef.png">

Wait what ? it is the same if I clean the Poland flag : 
![image](https://user-images.githubusercontent.com/5102622/204289661-e6cd5f72-a838-40ef-a546-54532a895e0a.png)

<img width="159" alt="image" src="https://user-images.githubusercontent.com/5102622/204289704-8851a738-106e-4ea7-8ba2-868739478ca0.png">

![image](https://user-images.githubusercontent.com/5102622/204289805-4f2c059a-b99b-4622-91ed-6d452b576656.png)

Then why is there so much useless attributes ? I don't know

I'll come back with a more radical way to clean the sprite

**Comment by thibault-mahe** [comment]
I open the draft of a PR here: https://github.com/Decathlon/vitamin-web/pull/1315 
But I have to test it a lot since it completely change the generated assets sprite and could bring breaking changes.

If you have some time I would like you to test the PR on your side if you can ;)

Thanks!

**Comment by thibault-mahe** [comment]
Hi @Jerome1337,

I tried all sort of fixes and nothing seems to work on my side. You can see part of these fixes on the  PR https://github.com/Decathlon/vitamin-web/pull/1315. The issue is that a large number of flags are dependent on `url()` (linear gradient, clip-path, ...). This feature is not well supported in browsers when used in combination with `<use>`. I listed in the PR a number of opened issue and stack overflow messages showing this issue is still opened.

I'll go back to my team and discuss with them if we can reexport all the flags with a simpler structure, and maybe discuss with them if we stop suggesting using svg with the `<use>` element...

in the meantime, I can suggest you 2 "not-so-great" fixes:

**First, you can switch method and use an image tag inseatd of the `<use>` element.**
Something like this I guess:
```
<img
  src={require(`../node_modules/@vtmn/assets/dist/assets/svg/${flag}.svg`)}
  alt=""
  aria-hidden="true"
  width="32"
  height="32"
/>
```

**Second, since the problem is related to the use of `url()` in the shadow DOM created by the `<use>` element, we can try to dynamically replace this `<use> `element by the svg directly inserted in the DOM :** 
```
const SPRITE_STORAGE_KEY = "VTMN_SPRITE";

// This function is responsible for parsing the fetched sprite, getting the wanted inner symbol
// and transforming it into a new svg element that will be inserted
const getSvgFromSprite = ({
  sprite,
  svgId,
}: {
  sprite: string;
  svgId: string;
}) => {
  const correspondingSymbol = new DOMParser()
    .parseFromString(sprite, "application/xml")
    .querySelector(`symbol${svgId}`);
  const correspondingViewbox = correspondingSymbol?.getAttribute("viewBox");

  return (
    <svg
      height="100%"
      width="100%"
      viewBox={correspondingViewbox as string}
      aria-hidden="true"
      dangerouslySetInnerHTML={{
        __html: correspondingSymbol?.innerHTML as string,
      }}
    ></svg>
  );
};

// This component is made to replace the SVG <use> element so the wanted svg is not inserted
// into a shadow DOM but directly in our DOM
const Use = ({ href }: { href: string }) => {
  const [use, setUse] = React.useState(<></>);
  const svgId = `#${href.split("#")[1]}`;

  React.useEffect(() => {
    const existingSprite = localStorage.getItem(SPRITE_STORAGE_KEY);
    if (existingSprite) {
      const svg = getSvgFromSprite({ sprite: existingSprite, svgId });
      setUse(svg);
    } else {
      fetch(href)
        .then((response) => response.text())
        .then((sprite) => {
          localStorage.setItem(SPRITE_STORAGE_KEY, sprite);
          const svg = getSvgFromSprite({ sprite, svgId });
          setUse(svg);
        });
    }
  }, []);

  return use;
};
```

Note that : 
- this method could largely be optimized, for instance you may want to clean your fetched sprite before using `dangerouslySetInnerHTML`, and offer better typing with interfaces, ...
- since the whole sprite would be fetched for each svg, you may want to cache it the first time you fetch it. For the purpose of the POC I used the local storage here, this is not a good idea in production. You may prefer caching him in the user RAM instead, like in a context or a shared class.

Does one of theses solutions could suit you eventually ?

Thanks

**Comment by Jerome1337** [comment]
Hi @thibault-mahe,

I think your first alternative should fit. I will try when I have some time.
I will switch back to the sprite method when it will be fixed and stable 👍 

I can test the PR in parallel and give a feedback too.

**Comment by thibault-mahe** [comment]
Ok thanks @Jerome1337 
Note that in the 1st solution I suggest there probably is a better way to get your icon than with an inlined require.
No need to test the PR, the suggested fix could not work in the current state of browsers unfortunately.

@lauthieb I'm leaving this issue opened since it is something we should discuss and address in a way or another: reworking all the flags or removing the sprite strategy.

Thanks!

**Comment by Jerome1337** [comment]
@thibault-mahe Everything is ok using the inline require method!
<img width="197" alt="Screenshot 2022-12-13 at 12 21 11 PM" src="https://user-images.githubusercontent.com/7968649/207304813-fdacc812-a767-4764-82f5-6223535472cd.png">

I documented everything on our side, we will fallback to the sprite method when it will be fixed and browsers states stables

**Comment by lauthieb** [comment]
Hello,

We just spent a few hours with the designers of the core team Design System to understand the problem of using sprite with our Vitamin assets (flags, payments, etc.).
In order to resolve it, we need to check all our different flags to remove masks & clip paths manually, and unfortunately, this is huge work & time-consuming which in regard to our team and the different priorities that we have, does not allow us to solve this soon.

That's why we took the decision to postpone the refactoring of all our flags in our next major release in design (@Decathlon/design-system-core-team-design), and so, we decided to no longer support the "sprite" consumption mode for `@vtmn/assets`.

So this issue had been closed with #1352.

Thanks for your understanding.
